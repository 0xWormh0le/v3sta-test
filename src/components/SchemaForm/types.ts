export type MoneyValidation = { maxValue?: number; minValue?: number }

export type StringValidation = { regex?: string }

export interface Schema {
  entity: 'Loan' | 'Borrower'
  display: string
  field: string
  type: 'string' | 'money' | 'date'
  conditions?: MoneyValidation | StringValidation
}

export type Entity = {
  [key: string]: string | Money | Date
}

export type StoreData = {
  Loan?: Entity
  Borrower?: Entity
}

export type Money = number | null

export type Date = { month: number; day: number; year: number } | null
