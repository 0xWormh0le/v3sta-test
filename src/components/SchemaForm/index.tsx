import React, { useCallback } from 'react'
import { FieldError, ChangeHandler, FieldValues, useForm } from 'react-hook-form'
import type { Schema } from './types'
import validateOption from './validate'

export * from './serialize'
export * from './types'

interface SchemaFormProps {
  schema: Schema[]
  title: string
  saveToStore?: (values: FieldValues) => void
  defaultValues?: { [key: string]: string }
  className?: string
}

const SchemaForm: React.FC<SchemaFormProps> = ({
  schema,
  title,
  saveToStore,
  defaultValues,
  className
}) => {
  const {
    register,
    getValues,
    getFieldState,
    formState: { errors }
  } = useForm({ mode: 'onTouched', defaultValues })

  const handleInputBlur = useCallback(
    (onBlur: ChangeHandler, name: string): React.FocusEventHandler<HTMLInputElement> =>
      (e) => {
        // need to execute onBlur passed from react hook form to trigger validation
        onBlur(e).then(() => {
          const { error } = getFieldState(name)
          if (error === undefined && saveToStore) {
            const values = getValues()
            // save only validated values
            const validValues = Object.keys(values)
              .filter((field) => errors[field] === undefined)
              .reduce((validValues, field) => ({ ...validValues, [field]: values[field] }), [])
            saveToStore(validValues)
          }
        })
      },
    [getValues, saveToStore, getFieldState, errors]
  )

  return (
    <form aria-label="schema-form" className={className}>
      <p className="text-2xl mb-6">{title}</p>
      {schema.map((fieldSchema, key) => {
        const { onBlur, ...other } = register(fieldSchema.field, validateOption(fieldSchema))
        return (
          <div key={key} className="mt-2">
            <label htmlFor={fieldSchema.field} className="block">
              {fieldSchema.display}
            </label>
            <input
              {...other}
              onBlur={handleInputBlur(onBlur, other.name)}
              id={fieldSchema.field}
              data-testid={fieldSchema.field}
              type={fieldSchema.type === 'date' ? 'date' : 'text'}
              className="p-1 w-full"
            />
            {errors[fieldSchema.field] && (
              <small className="text-red-600 block" data-testid={`error-${fieldSchema.field}`}>
                {(errors[fieldSchema.field] as FieldError).message}
              </small>
            )}
          </div>
        )
      })}
    </form>
  )
}

export default SchemaForm
