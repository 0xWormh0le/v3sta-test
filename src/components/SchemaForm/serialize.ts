import { FieldValues } from 'react-hook-form'
import type { Schema, StoreData, Money, Date } from './types'

/**
 * Parse money field of store and its return value is used in initial value of the form
 * @param value
 * @returns parsed value
 */
const parseMoney = (value: Money) => (value === null || value === undefined ? '' : value.toString())

/**
 * Parse date field of store and its return value is used in initial value of the form
 * @param value
 * @returns parsed value
 */
const parseDate = (value: Date) =>
  value === null || value === undefined
    ? ''
    : new Date(`${value.year}-${value.month}-${value.day}`).toISOString().slice(0, 10)

/**
 * Parse string field of store and its return value is used in initial value of the form
 * @param value
 * @returns parsed value
 */
const parseString = (value: string) => (value === undefined ? '' : value.toString())

/**
 * Format money field from the form and its return value will be saved in store after passing validation
 * @param value
 * @returns formated value
 */
export const formatMoney = (value: string) => {
  if (value === '') {
    return null
  } else if (parseFloat(value).toString() === value.toString()) {
    return Number(value)
  } else {
    return NaN
  }
}

/**
 * Format date field from the form and its return value will be saved in store
 * @param value
 * @returns formated value
 */
export const formatDate = (value: string) => {
  if (value === '') {
    return null
  } else {
    return {
      month: Number(value.slice(5, 7)),
      day: Number(value.slice(8)),
      year: Number(value.slice(0, 4))
    }
  }
}

/**
 * Take form value first, and current store value, and default blank value
 * @param formValue form value
 * @param currentStoreValue current store value
 * @param schema field schema
 * @returns final field value to be saved in store
 */
const pick = (formValue: any, currentStoreValue: any, schema: Schema) => {
  if (formValue !== undefined) {
    return formValue
  } else if (currentStoreValue !== undefined) {
    return currentStoreValue
  } else if (schema.type === 'string') {
    return ''
  } else {
    return null
  }
}

/**
 * Make value to be saved in store by merging current store data and form field values.
 *
 * If form value is undefined, it means the field is not validated from the form, hence won't be saved in the store.
 *
 * @param values form field values
 *
 * Each form field value can be:
 *
 * null: money or date is empty
 *
 * undefined: input is invalid, won't be saved in the store
 * @param currentStoreData current store data
 * @param schema form schema
 * @returns serialized value
 */
export const serialize = (
  values: FieldValues,
  currentStoreData: FieldValues | undefined,
  schema: Schema[]
) => {
  const data = {} as StoreData

  schema.forEach((s) => {
    const entity = data[s.entity]
    const value = pick(
      values[s.field],
      currentStoreData === undefined ? undefined : currentStoreData[s.field],
      s
    )

    if (entity === undefined) {
      data[s.entity] = { [s.field]: value }
    } else {
      entity[s.field] = value
    }
  })

  return JSON.stringify(data)
}

/**
 * Make value to be used as initial value of the form from the store data
 * 
 * It fills all missing fields with default value (null or empty string)
 *
 * Return `undefined` will be used as blank initial value
 * @note It takes care of any incorrect data from the store
 * @param rawData store data
 * @param schema form schema
 * @returns initial value of the form
 */
export const deserialize = (rawData: string, schema: Schema[]) => {
  try {
    const data: StoreData = JSON.parse(rawData)
    const value = {} as { [key: string]: string }

    schema.forEach((s) => {
      const entity = data[s.entity]

      if (entity === undefined) {
        value[s.field] = ''
      } else {
        const val = entity[s.field]
        try {
          switch (s.type) {
            case 'date':
              value[s.field] = parseDate(val as Date)
              return
            case 'money':
              value[s.field] = parseMoney(val as Money)
              return
            case 'string':
            default:
              value[s.field] = parseString(val as string)
              return
          }
        } catch {
          value[s.field] = ''
        }
      }
    })

    return value
  } catch {
    console.error('store data is not json format')
    return undefined
  }
}
