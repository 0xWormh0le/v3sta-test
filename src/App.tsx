import { FieldValues } from 'react-hook-form'
import SchemaForm, { Schema, serialize, deserialize } from 'components/SchemaForm'
import config from 'assets/schema.json'

const schema = config as Schema[]

const formKey = 'form-data'

export const saveToStore = (values: FieldValues) => {
  const currentStoreData = loadStore()
  const data = serialize(values, currentStoreData, schema)
  localStorage.setItem(formKey, data)
  // @todo api call here with data
  console.log(data)
}

export const loadStore = () => {
  const rawData = localStorage.getItem(formKey)
  if (rawData === null) {
    return undefined
  } else {
    return deserialize(rawData, schema)
  }
}

const App = () => (
  <SchemaForm
    schema={schema}
    title="Loan Information"
    saveToStore={saveToStore}
    defaultValues={loadStore()}
    className="mx-auto mt-12 p-3 w-full max-w-sm"
  />
)

export default App
