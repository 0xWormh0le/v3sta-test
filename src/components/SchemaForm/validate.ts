import { RegisterOptions } from 'react-hook-form'
import type { Schema, StringValidation, MoneyValidation } from './types'
import { formatMoney, formatDate } from './serialize'

/**
 * Gets validation option for the type of schema field
 * @param fieldSchema schema of the field
 * @returns validation option to be used in `register` function of react-hook-forms {@link https://react-hook-form.com/api/useform/register#options}
 */
const rule = (fieldSchema: Schema): RegisterOptions => {
  const rule: RegisterOptions = {}

  if (fieldSchema.type === 'string') {
    if (!fieldSchema.conditions) {
      return rule
    }

    const conditions = fieldSchema.conditions as StringValidation

    if (conditions.regex) {
      rule.pattern = {
        value: new RegExp(conditions.regex),
        message: 'Pattern mismatch'
      }
    }
  } else if (fieldSchema.type === 'money') {
    rule.setValueAs = formatMoney
    rule.validate = (value) => (isNaN(value) ? 'Number required' : true)

    if (!fieldSchema.conditions) {
      return rule
    }

    const conditions = fieldSchema.conditions as MoneyValidation

    if (conditions.minValue !== undefined) {
      rule.min = {
        value: conditions.minValue,
        message: `Must be no less than $${conditions.minValue}`
      }
    }

    if (conditions.maxValue !== undefined) {
      rule.max = {
        value: conditions.maxValue,
        message: `Must be no greater than $${conditions.maxValue}`
      }
    }
  } else {
    // date
    rule.setValueAs = formatDate
  }

  return rule
}

export default rule
