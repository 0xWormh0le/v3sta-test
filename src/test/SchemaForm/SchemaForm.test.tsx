import renderer from 'react-test-renderer'
import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import SchemaForm, { StoreData, Schema, Entity } from 'components/SchemaForm'
import App, { saveToStore, loadStore } from 'App'
import MockLocalStorage from 'test/mocks/LocalStorage'
import schemaNoValidation from 'test/schema/test.json'
import config from 'assets/schema.json'

const schema = config as Schema[]

const mockLocalStorage = new MockLocalStorage()

Object.defineProperty(global, 'localStorage', {
  value: mockLocalStorage
})

const log = jest.spyOn(console, 'log')
jest.spyOn(console, 'error')

const inputAndBlur = (fieldId: string, value: string) => {
  const input = screen.getByTestId(fieldId)

  input.focus()
  fireEvent.change(input, {
    target: { value }
  })
  input.blur()
}

describe('SchemaForm', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON()
    render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)

    expect(screen.getByTestId('loanAmount')).toBeInTheDocument()
    expect(screen.getByTestId('downPaymentAmount')).toBeInTheDocument()
    expect(screen.getByTestId('firstName')).toBeInTheDocument()
    expect(screen.getByTestId('lastName')).toBeInTheDocument()
    expect(screen.getByTestId('birthDate')).toBeInTheDocument()

    expect(tree).toMatchSnapshot('initial render')
  })

  describe('validation', () => {
    it('money validation fails: not number', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)
      inputAndBlur('loanAmount', 'text is not number')
      await waitFor(() => {
        expect(screen.getByTestId('error-loanAmount')).toHaveTextContent('Number required')
      })
      expect(log).not.toHaveBeenCalled()
    })

    it('money validation fails: min', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)
      inputAndBlur('loanAmount', '123')
      await waitFor(() => {
        expect(screen.getByTestId('error-loanAmount')).toHaveTextContent('Must be no less than')
      })
      expect(log).not.toHaveBeenCalled()
    })

    it('money validation fails: max', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)
      inputAndBlur('loanAmount', '99999999999')
      await waitFor(() => {
        expect(screen.getByTestId('error-loanAmount')).toHaveTextContent('Must be no greater than')
      })
      expect(log).not.toHaveBeenCalled()
    })

    it('string validation fails: pattern mismatch', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)
      inputAndBlur('firstName', 'invalid pattern 123')
      await waitFor(() => {
        expect(screen.getByTestId('error-firstName')).toHaveTextContent('Pattern mismatch')
      })
      expect(log).not.toHaveBeenCalled()
    })

    it('schema with optional validation', async () => {
      render(
        <SchemaForm
          title="Sample Form"
          schema={schemaNoValidation as Schema[]}
          saveToStore={saveToStore}
        />
      )
      inputAndBlur('firstName', 'number 123 is acceptable')
      inputAndBlur('loanAmount', '123')
      await waitFor(() => {
        expect(log).toHaveBeenCalled()
      })
      expect(screen.queryByTestId('error-firstName')).not.toBeInTheDocument()
    })
  })

  describe('save to store', () => {
    it('empty money and date will save to null', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)

      inputAndBlur('loanAmount', '')
      inputAndBlur('birthDate', '')

      await waitFor(() => {
        expect(log).toHaveBeenCalled()
      })

      const data = JSON.parse(localStorage.getItem('form-data') as string)

      expect(data.Loan.loanAmount).toBe(null)
      expect(data.Borrower.birthDate).toBe(null)
    })

    it('save only validated inputs and store will keep original value for unvalidated inputs', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)

      localStorage.setItem('form-data', JSON.stringify({ Loan: { loanAmount: '11111' }}))

      inputAndBlur('loanAmount', 'invalid string input')

      await waitFor(() => {
        expect(screen.getByTestId('error-loanAmount')).toHaveTextContent('Number required')
      })

      inputAndBlur('firstName', 'foo')

      await waitFor(() => {
        expect(log).toHaveBeenCalledTimes(1)
      })

      const data: StoreData = JSON.parse(localStorage.getItem('form-data') as string)

      expect((data.Loan as Entity).loanAmount).toBe('11111')
      expect((data.Borrower as Entity).firstName).toBe('foo')
    })

    it('save only validated inputs and unvalidated inputs will be saved as blank if the store doesn\'t include them', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)

      localStorage.clear()

      inputAndBlur('firstName', 'invalid name 123')

      await waitFor(() => {
        expect(screen.getByTestId('error-firstName')).toHaveTextContent('Pattern mismatch')
      })

      inputAndBlur('loanAmount', '111111')

      await waitFor(() => {
        expect(log).toHaveBeenCalledTimes(1)
      })

      const data = JSON.parse(localStorage.getItem('form-data') as string)

      expect((data.Loan as Entity).loanAmount).toBe(111111)
      expect((data.Loan as Entity).downPaymentAmount).toBe(null)
    })

    it('save to local storage on blur', async () => {
      render(<SchemaForm title="Sample Form" schema={schema} saveToStore={saveToStore} />)

      inputAndBlur('loanAmount', '11111')
      inputAndBlur('firstName', 'foo')
      inputAndBlur('lastName', 'bar')
      inputAndBlur('birthDate', '2022-12-11')

      await waitFor(() => {
        expect(log).toHaveBeenCalled()
      })

      const data: StoreData = JSON.parse(localStorage.getItem('form-data') as string)

      expect((data.Loan as Entity).loanAmount).toBe(11111)
      expect((data.Loan as Entity).downPaymentAmount).toBe(null)
      expect((data.Borrower as Entity).firstName).toBe('foo')
      expect((data.Borrower as Entity).lastName).toBe('bar')
      expect((data.Borrower as Entity).birthDate).toEqual({
        day: 11,
        month: 12,
        year: 2022
      })
    })
  })

  describe('load from store', () => {
    it('load from local storage', async () => {
      render(
        <SchemaForm
          title="Sample Form"
          schema={schema}
          saveToStore={saveToStore}
          defaultValues={loadStore()}
        />
      )

      expect(screen.getByTestId('loanAmount')).toHaveValue('11111')
      expect(screen.getByTestId('downPaymentAmount')).toHaveValue('')
      expect(screen.getByTestId('firstName')).toHaveValue('foo')
      expect(screen.getByTestId('lastName')).toHaveValue('bar')
      expect(screen.getByTestId('birthDate')).toHaveValue('2022-12-11')
    })

    it('load from empty local storage', async () => {
      localStorage.clear()

      render(
        <SchemaForm
          title="Sample Form"
          schema={schema}
          saveToStore={saveToStore}
          defaultValues={loadStore()}
        />
      )

      expect(screen.getByTestId('loanAmount')).toHaveValue('')
      expect(screen.getByTestId('downPaymentAmount')).toHaveValue('')
      expect(screen.getByTestId('firstName')).toHaveValue('')
      expect(screen.getByTestId('lastName')).toHaveValue('')
      expect(screen.getByTestId('birthDate')).toHaveValue('')
    })

    it('load from local storage with some missing fields', async () => {
      localStorage.setItem(
        'form-data',
        JSON.stringify({
          Loan: { loanAmount: 200000 },
          Borrower: { lastName: 'Jon' }
        })
      )

      render(
        <SchemaForm
          title="Sample Form"
          schema={schema}
          saveToStore={saveToStore}
          defaultValues={loadStore()}
        />
      )

      expect(screen.getByTestId('loanAmount')).toHaveValue('200000')
      expect(screen.getByTestId('downPaymentAmount')).toHaveValue('')
      expect(screen.getByTestId('firstName')).toHaveValue('')
      expect(screen.getByTestId('lastName')).toHaveValue('Jon')
      expect(screen.getByTestId('birthDate')).toHaveValue('')
    })

    it('load from local storage with missing entity', async () => {
      localStorage.setItem(
        'form-data',
        JSON.stringify({
          Borrower: { lastName: 'Jon' }
        })
      )

      render(
        <SchemaForm
          title="Sample Form"
          schema={schema}
          saveToStore={saveToStore}
          defaultValues={loadStore()}
        />
      )

      expect(screen.getByTestId('loanAmount')).toHaveValue('')
      expect(screen.getByTestId('downPaymentAmount')).toHaveValue('')
      expect(screen.getByTestId('firstName')).toHaveValue('')
      expect(screen.getByTestId('lastName')).toHaveValue('Jon')
      expect(screen.getByTestId('birthDate')).toHaveValue('')
    })

    it('store data is not json format', async () => {
      localStorage.setItem('form-data', 'non-json format')

      render(
        <SchemaForm
          title="Sample Form"
          schema={schema}
          saveToStore={saveToStore}
          defaultValues={loadStore()}
        />
      )

      expect(screen.getByTestId('loanAmount')).toHaveValue('')
      expect(screen.getByTestId('downPaymentAmount')).toHaveValue('')
      expect(screen.getByTestId('firstName')).toHaveValue('')
      expect(screen.getByTestId('lastName')).toHaveValue('')
      expect(screen.getByTestId('birthDate')).toHaveValue('')
    })

    it('keep all the rest value when a field is invalid', async () => {
      localStorage.setItem(
        'form-data',
        JSON.stringify({
          Borrower: { birthDate: 'non-date-format', firstName: 'Jon' },
          Loan: { loanAmount: 123456 }
        })
      )

      render(
        <SchemaForm
          title="Sample Form"
          schema={schema}
          saveToStore={saveToStore}
          defaultValues={loadStore()}
        />
      )

      expect(screen.getByTestId('loanAmount')).toHaveValue('123456')
      expect(screen.getByTestId('downPaymentAmount')).toHaveValue('')
      expect(screen.getByTestId('firstName')).toHaveValue('Jon')
      expect(screen.getByTestId('lastName')).toHaveValue('')
      expect(screen.getByTestId('birthDate')).toHaveValue('')
    })
  })
})
