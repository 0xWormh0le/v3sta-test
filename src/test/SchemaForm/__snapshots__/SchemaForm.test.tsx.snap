// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaForm renders correctly: initial render 1`] = `
<form
  aria-label="schema-form"
  className="mx-auto mt-12 p-3 w-full max-w-sm"
>
  <p
    className="text-2xl mb-6"
  >
    Loan Information
  </p>
  <div
    className="mt-2"
  >
    <label
      className="block"
      htmlFor="loanAmount"
    >
      Loan Amount
    </label>
    <input
      className="p-1 w-full"
      data-testid="loanAmount"
      id="loanAmount"
      name="loanAmount"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="mt-2"
  >
    <label
      className="block"
      htmlFor="downPaymentAmount"
    >
      Down Payment
    </label>
    <input
      className="p-1 w-full"
      data-testid="downPaymentAmount"
      id="downPaymentAmount"
      name="downPaymentAmount"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="mt-2"
  >
    <label
      className="block"
      htmlFor="firstName"
    >
      First Name
    </label>
    <input
      className="p-1 w-full"
      data-testid="firstName"
      id="firstName"
      name="firstName"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="mt-2"
  >
    <label
      className="block"
      htmlFor="lastName"
    >
      Last Name
    </label>
    <input
      className="p-1 w-full"
      data-testid="lastName"
      id="lastName"
      name="lastName"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="mt-2"
  >
    <label
      className="block"
      htmlFor="birthDate"
    >
      Birth Date
    </label>
    <input
      className="p-1 w-full"
      data-testid="birthDate"
      id="birthDate"
      name="birthDate"
      onBlur={[Function]}
      onChange={[Function]}
      type="date"
    />
  </div>
</form>
`;
